File: src/main/scala/com/hoi4utils/clausewitz/script/DynamicPDX.scala
package com.hoi4utils.clausewitz.script

import com.hoi4utils.clausewitz_parser.Node
import org.jetbrains.annotations.NotNull
import org.jetbrains.annotations.Nullable

import java.util.function.Supplier

import scala.collection.mutable.ListBuffer


class DynamicPDX[V, U <: StructuredPDX] extends PDXScript[V] {
  protected var simplePDX: PDXScript[V] = _
  protected var simplePDXSupplier: () => PDXScript[V] = _
  final protected var structuredBlock: U = null.asInstanceOf[U]
  /**
   * may be null. the structured block does not always have a single property
   * that is equivalent to the simple value. Or, the structured block has a
   * property that is equivalent to a value and the property is null.
   */
  private var structuredPDXValueIdentifiers: List[String] = _

  def this(simplePDXSupplier: () => PDXScript[V], structuredBlock: U) = {
    this()
    this.simplePDXSupplier = simplePDXSupplier
    this.structuredBlock = structuredBlock
    this.structuredPDXValueIdentifiers = null
  }

  def this(simplePDXSupplier: () => PDXScript[V], structuredBlock: U, structuredPDXValueIdentifier: String) = {
    this()
    this.simplePDXSupplier = simplePDXSupplier
    this.structuredBlock = structuredBlock
    this.structuredPDXValueIdentifiers = List(structuredPDXValueIdentifier)
  }

  def this(simplePDXSupplier: () => PDXScript[V], structuredBlock: U, structuredPDXValueIdentifier: List[String]) = {
    this()
    this.simplePDXSupplier = simplePDXSupplier
    this.structuredBlock = structuredBlock
    this.structuredPDXValueIdentifiers = structuredPDXValueIdentifier
  }

  def this(structuredBlock: U) = {
    this()
    this.simplePDXSupplier = null
    this.structuredBlock = structuredBlock
    this.structuredPDXValueIdentifiers = null
  }

  def this(structuredBlock: U, structuredPDXValueIdentifier: String) = {
    this()
    this.simplePDXSupplier = null
    this.structuredBlock = structuredBlock
    this.structuredPDXValueIdentifiers = List(structuredPDXValueIdentifier)
  }

  def this(structuredBlock: U, structuredPDXValueIdentifier: List[String]) = {
    this()
    this.simplePDXSupplier = null
    this.structuredBlock = structuredBlock
    this.structuredPDXValueIdentifiers = structuredPDXValueIdentifier
  }

  def this(simplePDXSupplier: () => PDXScript[V]) = {
    this()
    this.simplePDXSupplier = simplePDXSupplier
    this.structuredBlock = null.asInstanceOf[U]
    this.structuredPDXValueIdentifiers = null
  }

  def this(simplePDXSupplier: () => PDXScript[V], structuredPDXValueIdentifier: String) = {
    this()
    this.simplePDXSupplier = simplePDXSupplier
    this.structuredBlock = null.asInstanceOf[U]
    this.structuredPDXValueIdentifiers = List(structuredPDXValueIdentifier)
  }

  def this(simplePDXSupplier: () => PDXScript[V], structuredPDXValueIdentifier: List[String]) = {
    this()
    this.simplePDXSupplier = simplePDXSupplier
    this.structuredBlock = null.asInstanceOf[U]
    this.structuredPDXValueIdentifiers = structuredPDXValueIdentifier
  }

  def setSimplePDX(simplePDX: PDXScript[V]): Unit = {
    this.simplePDX = simplePDX
    this.setNull(structuredBlock)
  }

  private def setNull(pdx: AbstractPDX[?]): Unit = {
    if (pdx != null) pdx.setNull()
  }

  def setSimplePDX(obj: V): Unit = {
    if (obj == null) {
      simplePDX = null
      return
    }
    supplySimplePDX()
    simplePDX.set(obj)
  }

  def isBlock: Boolean = simplePDX == null

  private def isBlock(expression: Node) = {
    expression.$ match {
      case l: ListBuffer[Node] => true
      case _ => false
    }
  }

  override def set(obj: V): Unit = {
    if (!isBlock) setSimplePDX(obj)
  }

  @throws[UnexpectedIdentifierException]
  @throws[NodeValueTypeException]
  def set(expression: Node): Unit = {
    if (!isBlock(expression)) {
      supplySimplePDX()
      simplePDX.set(expression)
    }
    else if (blockAllowed) structuredBlock.set(expression)
  }

  override def get(): V = if (!isBlock) simplePDX.get()
  else if (isBlock) {
    val valueProperty = getStructuredValueProperty
    if (valueProperty == null) null.asInstanceOf[V]
    else valueProperty.get()
  }
  else null.asInstanceOf[V]

  @throws[UnexpectedIdentifierException]
  def loadPDX(expression: Node): Unit = {
    if (!isBlock(expression)) {
      supplySimplePDX()
      simplePDX.loadPDX(expression)
    }
    else if (blockAllowed) structuredBlock.loadPDX(expression)
  }

  private def supplySimplePDX(): Unit = {
    if (simplePDX == null) simplePDX = simplePDXSupplier()
  }

  def loadPDX(expressions: ListBuffer[Node]): Unit = {
    if (blockAllowed) structuredBlock.loadPDX(expressions)
    simplePDX.setNull()
  }

  def isValidIdentifier(node: Node): Boolean = {
    // todo i hope this works as intended
    if (!isBlock) simplePDX.isValidIdentifier(node)
    else if (blockAllowed) structuredBlock.isValidIdentifier(node)
    else false
  }

  private def blockAllowed = structuredBlock != null

  override def setNull(): Unit = {
    simplePDX = null
    setNull(structuredBlock)
  }

  def loadOrElse(exp: Node, value: V): Unit = {
    if (!isBlock) simplePDX.loadOrElse(exp, value)
    else if (blockAllowed) structuredBlock.loadOrElse(exp, null)
  }

  override def toScript: String = if (!isBlock) simplePDX.toScript
  else if (blockAllowed) structuredBlock.toScript
  else "[null DynamicPDX]"

  override def nodeEquals(other: PDXScript[?]): Boolean = {
    if (!isBlock) simplePDX.equals(other)
    else if (blockAllowed) structuredBlock.equals(other)
    else false
  }

  override def getOrElse(elseValue: V): V = {
    get() match {
      case null => elseValue
      case v: V => v
      case _ => throw new NodeValueTypeException(null, "V")
    }
  }

  override def isUndefined: Boolean = get() == null

  override def toString: String = "DynamicPDX{" + "simpleValue=" + simplePDX + ", structuredValue=" + structuredBlock + '}'

  /**
   * may be null. the structured block does not always have a single property
   * that is equivalent to the simple value. Or, the structured block has a
   * property that is equivalent to a value and the property is null.
   */
  private def getStructuredValueProperty = {
    if (isBlock && blockAllowed && structuredPDXValueIdentifiers != null) structuredBlock.getPDXPropertyOfType[V](structuredPDXValueIdentifiers)
    else null
  }

  override def getPDXIdentifier: String = if (!isBlock) simplePDX.getPDXIdentifier
  else if (blockAllowed) structuredBlock.getPDXIdentifier
  else null

  def getPDXScript: PDXScript[?] = if (!isBlock) simplePDX
  else structuredBlock
}

File: src/main/java/com/hoi4utils/clausewitz/script/DynamicPDX.java
package com.hoi4utils.clausewitz.script;

import com.hoi4utils.clausewitz_parser.Node;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.function.Supplier;

public abstract class DynamicPDX<V, U extends StructuredPDX> implements PDXScript<V> {
    protected PDXScript<V> simplePDX;
    protected Supplier<PDXScript<V>> simplePDXSupplier;
    @Nullable protected final U structuredBlock;
    /**
     * may be null. the structured block does not always have a single property
     * that is equivalent to the simple value. Or, the structured block has a
     * property that is equivalent to a value and the property is null.
     */
    private List<String> structuredPDXValueIdentifiers;

    public DynamicPDX(@NotNull Supplier<PDXScript<V>> simplePDXSupplier, @NotNull U structuredBlock) {
        this.simplePDXSupplier = simplePDXSupplier;
        this.structuredBlock = structuredBlock;
        this.structuredPDXValueIdentifiers = null;
    }

    public DynamicPDX(@NotNull Supplier<PDXScript<V>> simplePDXSupplier, @NotNull U structuredBlock, String structuredPDXValueIdentifier) {
        this.simplePDXSupplier = simplePDXSupplier;
        this.structuredBlock = structuredBlock;
        this.structuredPDXValueIdentifiers = List.of(structuredPDXValueIdentifier);
    }

    public DynamicPDX(@NotNull Supplier<PDXScript<V>> simplePDXSupplier, @NotNull U structuredBlock, List<String> structuredPDXValueIdentifier) {
        this.simplePDXSupplier = simplePDXSupplier;
        this.structuredBlock = structuredBlock;
        this.structuredPDXValueIdentifiers = structuredPDXValueIdentifier;
    }

    public DynamicPDX(@NotNull U structuredBlock) {
        this.simplePDXSupplier = null;
        this.structuredBlock = structuredBlock;
        this.structuredPDXValueIdentifiers = null;
    }

    public DynamicPDX(@NotNull U structuredBlock, String structuredPDXValueIdentifier) {
        this.simplePDXSupplier = null;
        this.structuredBlock = structuredBlock;
        this.structuredPDXValueIdentifiers = List.of(structuredPDXValueIdentifier);
    }

    public DynamicPDX(@NotNull U structuredBlock, List<String> structuredPDXValueIdentifier) {
        this.simplePDXSupplier = null;
        this.structuredBlock = structuredBlock;
        this.structuredPDXValueIdentifiers = structuredPDXValueIdentifier;
    }

    public DynamicPDX(@NotNull Supplier<PDXScript<V>> simplePDXSupplier) {
        this.simplePDXSupplier = simplePDXSupplier;
        this.structuredBlock = null;
        this.structuredPDXValueIdentifiers = null;
    }

    public DynamicPDX(@NotNull Supplier<PDXScript<V>> simplePDXSupplier, String structuredPDXValueIdentifier) {
        this.simplePDXSupplier = simplePDXSupplier;
        this.structuredBlock = null;
        this.structuredPDXValueIdentifiers = List.of(structuredPDXValueIdentifier);
    }

    public DynamicPDX(@NotNull Supplier<PDXScript<V>> simplePDXSupplier, List<String> structuredPDXValueIdentifier) {
        this.simplePDXSupplier = simplePDXSupplier;
        this.structuredBlock = null;
        this.structuredPDXValueIdentifiers = structuredPDXValueIdentifier;
    }

    public void setSimplePDX(PDXScript<V> simplePDX) {
        this.simplePDX = simplePDX;
        this.setNull(structuredBlock);
    }

    private void setNull(AbstractPDX<?> pdx) {
        if (pdx != null) {
            pdx.setNull();
        }
    }

    public void setSimplePDX(V obj) {
        if (obj == null) {
            simplePDX = null;
            return;
        }

        supplySimplePDX();
        simplePDX.set(obj);
    }

    public boolean isBlock() {
        return simplePDX == null;
    }

    private boolean isBlock(Node expression) {
        return expression.value().isList();
    }

    @Override
    public void set(V obj) {
        if (!isBlock()) {
            setSimplePDX(obj);
        } else {
            return;
        }
    }

    @Override
    public void set(Node expression) throws UnexpectedIdentifierException, NodeValueTypeException {
        if (!isBlock(expression)) {
            supplySimplePDX();
            simplePDX.set(expression);
        } else if (blockAllowed()) {
            structuredBlock.set(expression);
        }
    }

    @Override
    public @Nullable V get() {
        if (!isBlock()) {
            return simplePDX.get();
        } else if (isBlock()) {
            var valueProperty = getStructuredValueProperty();
            return valueProperty == null ? null : valueProperty.get();
        } else {
            return null;
        }
    }

    @Override
    public void loadPDX(Node expression) throws UnexpectedIdentifierException {
        if (!isBlock(expression)) {
            supplySimplePDX();
            simplePDX.loadPDX(expression);
        } else if (blockAllowed()) {
            structuredBlock.loadPDX(expression);
        }
    }

    private void supplySimplePDX() {
        if (simplePDX == null) {
            simplePDX = simplePDXSupplier.get();
        }
    }

    @Override
    public void loadPDX(List<Node> expressions) {
        if (blockAllowed())
            structuredBlock.loadPDX(expressions);
        simplePDX.setNull();
    }

    @Override
    public boolean isValidIdentifier(Node node) {
        // todo i hope this works as intended
        if (!isBlock()) {
            return simplePDX.isValidIdentifier(node);
        } else if (blockAllowed()) {
            return structuredBlock.isValidIdentifier(node);
        } else {
            return false;
        }
    }

    private boolean blockAllowed() {
        return structuredBlock != null;
    }

    @Override
    public void setNull() {
        simplePDX = null;
        setNull(structuredBlock);
    }

    @Override
    public void loadOrElse(Node exp, V value) {
        if (!isBlock()) {
            simplePDX.loadOrElse(exp, value);
        } else if (blockAllowed()) {
            structuredBlock.loadOrElse(exp, null);
//            setStructuredBlockValue(value);
        }
    }

    @Override
    public String toScript() {
        if (!isBlock()) {
            return simplePDX.toScript();
        } else if (blockAllowed()) {
            return structuredBlock.toScript();
        } else {
            return "[null DynamicPDX]";
        }
    }

    @Override
    public boolean objEquals(PDXScript<?> other) {
        if (!isBlock()) {
            return simplePDX.objEquals(other);
        } else if (blockAllowed()) {
            return structuredBlock.objEquals(other);
        } else {
            return false;
        }
    }

    @Override
    public V getOrElse(V elseValue) {
        V get = get();
        return get == null ? elseValue : get;
    }

    @Override
    public boolean isUndefined() {
        return get() == null;
    }

    @Override
    public String toString() {
        return "DynamicPDX{" +
                "simpleValue=" + simplePDX +
                ", structuredValue=" + structuredBlock +
                '}';
    }


    /**
     * may be null. the structured block does not always have a single property
     * that is equivalent to the simple value. Or, the structured block has a
     * property that is equivalent to a value and the property is null.
     */
    private @Nullable PDXScript<V> getStructuredValueProperty() {
        if (isBlock() && blockAllowed() && structuredPDXValueIdentifiers != null) {
            return structuredBlock.<V>getPDXPropertyOfType(structuredPDXValueIdentifiers);
        } else {
            return null;
        }
    }

    @Override
    public String getPDXIdentifier() {
        if (!isBlock()) {
            return simplePDX.getPDXIdentifier();
        } else if (blockAllowed()) {
            return structuredBlock.getPDXIdentifier();
        } else {
            return null;
        }
    }

    public PDXScript<?> getPDXScript() {
        if (!isBlock()) {
            return simplePDX;
        } else {
            return structuredBlock;
        }
    }
}
